onnxruntime/contrib_ops/cpu/tokenizer.cc

```diff
@@ -0,0 +1,472 @@
+      return get(node->left_.get(), str, len, depth);
+    } else if (c > node->c_) {
+      return get(node->right_.get(), str, len, depth);
+    } else if (depth < (len - 1)) {
+      if (node->mid_ != nullptr) {
+        return get(node->mid_.get(), str, len, depth + 1);
+      }
+    }
+    return node;
+  }
+
+  Node* put(Node* node, const CharT* str, size_t len, const Value& v, size_t depth) {
+    CharT c = str[depth];
+
+    std::unique_ptr<Node> new_node;
+    if (node == nullptr) {
+      new_node.reset(new Node(c));
+    }
+
+    Node* new_link = nullptr;
+    Node* n = (node != nullptr) ? node : new_node.get();
+    if (c < n->c_) {
+      new_link = put(n->left_.get(), str, len, v, depth);
+      if (new_link != nullptr) {
+        n->left_.release();
+        n->left_.reset(new_link);
+      }
+    } else if (c > n->c_) {
+      new_link = put(n->right_.get(), str, len, v, depth);
+      if (new_link != nullptr) {
+        n->right_.release();
+        n->right_.reset(new_link);
+      }
+    } else if (depth < (len - 1)) {
+      new_link = put(n->mid_.get(), str, len, v, depth + 1);
+      if (new_link != nullptr) {
+        n->mid_.release();
+        n->mid_.reset(new_link);
+      }
+    } else {
+      if (!n->has_val_) {
+        n->value_ = v;
+        n->has_val_ = true;
+        new_link = n;
+      }
+    }
+    if (new_link != nullptr) {
+      new_node.release();
+      return n;
+    }
+    return nullptr;
+  }
+  std::unique_ptr<Node> root_;
+};
+
+// We store the length of the original pattern within the
+// Ternary Tree so on search miss we drop this many characters from the
+// string
+struct SearchValue {
+  size_t w_len;
+  int priority_;
+};
+
+}  // namespace tokenizer_details
+
+using namespace tokenizer_details;
+
+struct Tokenizer::SearchData {
+  TernarySearchTree<wchar_t, SearchValue> tst_;
+};
+
+Tokenizer::Tokenizer(const OpKernelInfo& info) : OpKernel(info), search_data_(nullptr) {
+  int64_t mark = 0;
+  auto status = info.GetAttr("mark", &mark);
+  ONNXRUNTIME_ENFORCE(status.IsOK(), "attribute mark is not set");
+  mark_ = mark != 0;
+  status = info.GetAttr("pad_value", &pad_value_);
+  ONNXRUNTIME_ENFORCE(status.IsOK(), "attribute pad_value is not set");
+  status = info.GetAttr("mincharnum", &mincharnum_);
+  ONNXRUNTIME_ENFORCE(status.IsOK(), "attribute mincharnum is not set");
+
+  std::vector<std::string> separators;
+  status = info.GetAttrs<std::string>("separators", separators);
+  ONNXRUNTIME_ENFORCE(status.IsOK(), "attribute separators is not set");
+  ONNXRUNTIME_ENFORCE(!separators.empty(), "Requires at least one separator");
+
+  char_tokenezation_ = (separators.size() == 1 &&
+                        separators[0].empty());
+
+  ONNXRUNTIME_ENFORCE(!char_tokenezation_ || mincharnum_ < 2,
+                      "mincharnum is too big for char level tokenezation");
+
+  // Create TST and insert separators
+  if (!char_tokenezation_) {
+    std::unique_ptr<SearchData> sd(new SearchData);
```

Comment:
std::make_unique

Reply:
make_unique is not always present in GCC versions.

---
In reply to: [238895544](https://github.com/Microsoft/onnxruntime/pull/31#discussion_r238895544) [](ancestors = 238895544)

Reply:
Our code makes use of make_unique extensively and relies on it's existence. It's safe to assume it'll be there. I think we even have a function where we define our own make_unique when not present.